Support target-typed new expressions

anserg256 opened the pull request:
Hello, @angularsen!
This one addresses the issue with target-typed new expressions not being recognized as object initializers.

Closes #8

anserg256 committed:
Recognize ImplicitObjectCreationExpressionSyntax

anserg256 committed:
Add tests and sample for target-typed new

Pankraty commented on AssignAll/AssignAll/AssignAllMembers/ObjectInitializerAnalyzer.cs:
```diff
@@ -153,5 +155,15 @@ private static ImmutableArray<string> GetIgnoredPropertyNames(ObjectCreationExpr
+
+        private BaseObjectCreationExpressionSyntax GetObjectCreation(SyntaxNode syntaxNode)
+        {
+            BaseObjectCreationExpressionSyntax result = syntaxNode as ObjectCreationExpressionSyntax;
+
+            if (result == null)
+                result = syntaxNode as ImplicitObjectCreationExpressionSyntax;
+
+            return result;
```
Will it not be effectively the same as
```suggestion
            return syntaxNode as BaseObjectCreationExpressionSyntax;
```
(given there are only two classes derived from `BaseObjectCreationExpressionSyntax`, according to https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.syntax.baseobjectcreationexpressionsyntax?view=roslyn-dotnet-4.3.0).

anserg256 replied:
Right, there's no need to be that specific. Thanks!

anserg256 committed:
Simplify syntax node identification

anserg256 committed:
Return pattern matching

angularsen commented:
Thanks, I didn't have to time to look into it this weekend, but will get to it.

In the meantime, I did an attempt on this earlier but reverted it.
https://github.com/angularsen/roslyn-analyzers/commit/309d875690ab2ad2f5c04c13f34475e9df3c44d9

I don't recall exactly, but I believe I had to upgrade some nugets to make the new C# syntax work in the analyzer, and that resulted in breaking the plugin altogether and it all seemed to be due to the nuget upgrades.

I then tried to do the the upgrades separately and make sure that worked, before re-trying the `new()` support, but never got around to it.

[⬆️ Try upgrade 3.11.0 to to 4.1.0](https://github.com/angularsen/roslyn-analyzers/commit/33fa973e210ccff0b19b6e1b278d7e9667420e67)

angularsen committed:
Minor improvements cherry picked from 309d875

Cherry picked changes from 309d875690ab2ad2f5c04c13f34475e9df3c44d9

Minor improvements to syntax, null checking and unit test.

angularsen commented:
It took me some time to get back to this, sorry.
I was able to verify that a locally produced nuget does enable this new analyzer feature, so this should be good to go.

Nuget should be out shortly.

[Release 1.7.2 · angularsen/roslyn-analyzers](https://github.com/angularsen/roslyn-analyzers/releases/tag/1.7.2)
